### PR TITLE
Adding Support for Metadata

### PR DESCRIPTION
**Pull request by emarinier**
This is a basic first pass of adding support for metadata.

There are other things that need to be updated in the pipeline, but this PR tries to tackle the metadata-related things and only changes other things if there very important (although such related things might need another PR).

## Command

```
rm -rf results; nextflow run main.nf -profile docker --input tests/data/samplesheets/samplesheet.csv --outdir results -params-file assets/parameters.yaml
```

## Input

[samplesheet.csv](https://github.com/phac-nml/clustersplitter/blob/metadata/tests/data/samplesheets/samplesheet.csv): 
```
sample,mlst_alleles,metadata_1,metadata_2,metadata_3,metadata_4,metadata_5,metadata_6,metadata_7,metadata_8
S1,https://raw.githubusercontent.com/phac-nml/clustersplitter/dev/tests/data/profiles/S1.mlst.json,1,"Escherichia coli","EHEC/STEC","Canada","O157:H7",21,"2024/05/30","beef"
S2,https://raw.githubusercontent.com/phac-nml/clustersplitter/dev/tests/data/profiles/S2.mlst.json,1,"Escherichia coli","EHEC/STEC","The United States","O157:H7",55,"2024/05/21","milk"
S3,https://raw.githubusercontent.com/phac-nml/clustersplitter/dev/tests/data/profiles/S3.mlst.json,2,"Escherichia coli","EPEC","France","O125",14,"2024/04/30","cheese"
S4,https://raw.githubusercontent.com/phac-nml/clustersplitter/dev/tests/data/profiles/S4.mlst.json,2,"Escherichia coli","EPEC","France","O125",35,"2024/04/22","cheese"
S5,https://raw.githubusercontent.com/phac-nml/clustersplitter/dev/tests/data/profiles/S5.mlst.json,3,"Escherichia coli","EAEC","Canada","O126:H27",61,"2012/09/01","milk"
S6,https://raw.githubusercontent.com/phac-nml/clustersplitter/dev/tests/data/profiles/S6.mlst.json,unassociated,"Escherichia coli","EAEC","Canada","O111:H21",43,"2011/12/25","fruit"
```

[config.json](https://github.com/phac-nml/clustersplitter/blob/metadata/tests/data/configs/config.json)
```
    {
        "outlier_thresh": "25",
        "clustering_method": "average",
        "clustering_threshold": "500,100,75,50,25,15,10,5,2,1,0",
        "min_cluster_members": 2,
        "partition_column_name": "outbreak",
        "id_column_name": "sample_id",
        "only_report_labeled_columns": "False",
        "skip_qa": "False",
        
        "grouped_metadata_columns":{ 
            "outbreak":{"data_type": "None","label":"National Outbreak Code","default":"","display":"True"},
            "organism":{"data_type": "None","label":"Organism","default":"","display":"True"},
            "subtype":{"data_type": "None","label":"Subtype","default":"","display":"True"},
            "country":{"data_type": "categorical","label":"Country of Collection","default":"","display":"True"},
            "serovar":{"data_type": "categorical","label":"Serovar","default":"","display":"True"},
            "age":{"data_type": "desc_stats","label":"Patient Age (years)","default":"","display":"True"},
            "date":{"data_type": "min_max","label":"Date","default":"","display":"True"},
            "source":{"data_type": "categorical","label":"Source Type","default":"","display":"True"}
        },

        "linelist_columns":{
            "sample":{"data_type": "None","label":"Sample","default":"","display":"True"},
            "outbreak":{"data_type": "None","label":"National Outbreak Code","default":"","display":"True"},
            "organism":{"data_type": "None","label":"Organism","default":"","display":"True"},
            "subtype":{"data_type": "None","label":"Subtype","default":"","display":"True"},
            "country":{"data_type": "categorical","label":"Country of Collection","default":"","display":"True"},
            "serovar":{"data_type": "categorical","label":"Serovar","default":"","display":"True"},
            "age":{"data_type": "desc_stats","label":"Patient Age (years)","default":"","display":"True"},
            "date":{"data_type": "min_max","label":"Date","default":"","display":"True"},
            "source":{"data_type": "categorical","label":"Source Type","default":"","display":"True"}
        }    
    }
```

[parameters.yaml](https://github.com/phac-nml/clustersplitter/blob/metadata/assets/parameters.yaml)
```
partition_column: "outbreak"
metadata_1_header : "outbreak"
metadata_2_header : "organism"
metadata_3_header : "subtype"
metadata_4_header : "country"
metadata_5_header : "serovar"
metadata_6_header : "age"
metadata_7_header : "date"
metadata_8_header : "source"
```

## Output

cluster_summary.tsv
```
National Outbreak Code	Organism	Subtype	Country of Collection	Serovar	Patient Age (years)	Date	Source Type	date_max_value	count_age_43	count_age_21	count_serovar_O125	count_serovar_O157:H7	count_age_55	date_min_value	count_country_The United States	max_dist	count_age_14	count_outliers	count_source_fruit	count_subtype_EAEC	min_dist	age_max_value	count_source_beef	count_serovar_O126:H27	median_dist	count_subtype_EHEC/STEC	age_min_value	count_country_France	age_mean_value	count_serovar_O111:H21	mean_dist	count_organism_Escherichia coli	count_source_milk	count_subtype_EPEC	count_members	count_age_35	count_age_61	age_median_value	count_source_cheese	count_country_Canada
1	Escherichia coli	EHEC/STEC	Canada,The United States	O157:H7	21,55	2024/05/21,2024/05/30	beef,milk	55.0	0	0	0	2	0	21.0	1	3.0	0	0	0	0	3.0	55.0	1	0	3.0	0	21.0	0	38.0	0	3.0	0	1	0	2	0	0	38.0	0	1
2	Escherichia coli	EPEC	France	O125	14,35	2024/04/22,2024/04/30	cheese	35.0	0	0	2	0	0	14.0	0	2.0	0	0	0	0	2.0	35.0	0	0	2.0	0	14.0	2	24.5	0	2.0	0	0	0	2	0	0	24.5	2	0
3	Escherichia coli	EAEC	Canada	O126:H27	61	2012/09/01	milk	61.0	0	0	0	0	0	61.0	0	0	0	0	0	0	0	61.0	0	1	0	0	61.0	0	61.0	0	0	0	1	0	1	0	0	61.0	0	1
unassociated	Escherichia coli	EAEC	Canada	O111:H21	43	2011/12/25	fruit	43.0	0	0	0	0	0	43.0	0	0	0	0	1	0	0	43.0	0	0	0	0	43.0	0	43.0	1	0	0	0	0	1	0	0	43.0	0	1
```

metadata.included.tsv
```
sample	outbreak	organism	subtype	country	serovar	age	date	source
S1	1	Escherichia coli	EHEC/STEC	Canada	O157:H7	21	2024/05/30	beef
S2	1	Escherichia coli	EHEC/STEC	The United States	O157:H7	55	2024/05/21	milk
S3	2	Escherichia coli	EPEC	France	O125	14	2024/04/30	cheese
S4	2	Escherichia coli	EPEC	France	O125	35	2024/04/22	cheese
S5	3	Escherichia coli	EAEC	Canada	O126:H27	61	2012/09/01	milk
S6	unassociated	Escherichia coli	EAEC	Canada	O111:H21	43	2011/12/25	fruit
```

### CODE REVIEW

**Review by mattheww95 — APPROVED** [review]
Looks good to me!

**Review by kylacochrane — APPROVED** [review]
Everything looks good Eric 👍

**Review by apetkau — CHANGES_REQUESTED** [review]
Thanks so much @emarinier . This is great 😄 . Comments below

File: tests/data/arborview/expected_1.html [hunk 1]
```diff
@@ -110,7 +110,7 @@
...
         const TREE = "(S2:3.000000,S1:3.000000);\n";
         // the '= __DEADFOOD__' is an expression to find and replace and inline tsv string e.g. 'head1\thead2\nv1\tv2\n'
-        const DATA = "id\tmd_1\tmd_2\tmd_3\tmd_4\tmd_5\nS1\t1\t\t\t\t\nS2\t1\t\t\t\t\n";
+        const DATA = "sample\toutbreak\torganism\tsubtype\tcountry\tserovar\tage\tdate\tsource\nS1\t1\tEscherichia coli\tEHEC/STEC\tCanada\tO157:H7\t21\t2024/05/30\tbeef\nS2\t1\tEscherichia coli\tEHEC/STEC\tThe United States\tO157:H7\t55\t2024/05/21\tmilk\n";
```
[the remark this answers is not in the document]
Comment (emarinier) [reply tests/data/arborview/expected_1.html]: Updated in 02df5922f21bf09f86b41182d14d934e21bccabb

File: workflows/cluster_splitter.nf [hunk 2]
```diff
@@ -88,7 +96,7 @@ workflow CLUSTER_SPLITTER {
         metadata=merged_metadata,
         configuration_file=arborator_config,
         id_column=ID_COLUMN,
-        partition_col=PARTITION_COLUMN,
+        partition_col=params.partition_column,
```
[the remark this answers is not in the document]
Comment (emarinier) [reply workflows/cluster_splitter.nf]: Updated in 916ac01bd321980acefde078ee5c4b7ed8ff04a9

**Review by apetkau — APPROVED** [review]
This looks great to me. Thanks so much @emarinier 😄

**Pull request merged**